prov/gni/include/gnix_tags.h

```diff
@@ -0,0 +1,175 @@
+/*
+ * Copyright (c) 2015 Cray Inc. All rights reserved.
+ *
+ * This software is available to you under a choice of one of two
+ * licenses.  You may choose to be licensed under the terms of the GNU
+ * General Public License (GPL) Version 2, available from the file
+ * COPYING in the main directory of this source tree, or the
+ * BSD license below:
+ *
+ *     Redistribution and use in source and binary forms, with or
+ *     without modification, are permitted provided that the following
+ *     conditions are met:
+ *
+ *      - Redistributions of source code must retain the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer.
+ *
+ *      - Redistributions in binary form must reproduce the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer in the documentation and/or other materials
+ *        provided with the distribution.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#ifndef PROV_GNI_SRC_GNIX_TAGS_H_
+#define PROV_GNI_SRC_GNIX_TAGS_H_
+
+#include <fi.h>
+#include <fi_list.h>
+
+#include "gnix_util.h"
+
+/* enumerations */
+enum {
+	GNIX_TAG_AUTOSELECT = 0,
+	GNIX_TAG_LIST,
+	GNIX_TAG_HLIST,
+	GNIX_TAG_KDTREE,
+	GNIX_TAG_MAXTYPES,
+};
+
+enum {
+	GNIX_TS_STATE_UNINITIALIZED = 0,
+	GNIX_TS_STATE_INITIALIZED,
+	GNIX_TS_STATE_DESTROYED,
+};
+
+/* forward declarations */
+struct gnix_tag_storage;
+struct gnix_fab_req;
+
+/* structure declarations */
+struct gnix_tag_storage_ops {
+	int (*insert_tag)(struct gnix_tag_storage *ts, uint64_t tag,
+			struct gnix_fab_req *req);
+	struct gnix_fab_req *(*remove_tag)(struct gnix_tag_storage *ts,
+			uint64_t tag, uint64_t ignore);
+	struct gnix_fab_req *(*peek_tag)(struct gnix_tag_storage *ts,
+			uint64_t tag, uint64_t ignore);
+	int (*init)(struct gnix_tag_storage *ts);
+	int (*fini)(struct gnix_tag_storage *ts);
+};
+
+struct gnix_tag_list_element {
+	struct slist_entry free;
+	uint32_t seq;
+	uint32_t gen;
```

Comment:
The gen field protects against sequence wrap when using the 'seq' and 'gen' fields in the hlist and kdtree implementations. In the case of a wildcard, we may have many potential matches and the first match we find in a hlist or kdtree implementation may not be the oldest possible match. So using the sequence and generation fields allows us to retain some state about the search space to do a branch and bound approach. If one match is found, we save it as the oldest. Then, while checking other lists or nodes, if we find that the sequence we are currently inspecting is newer than the oldest found sequence, we prevent further searching of the lists. 

In a single list, it isn't necessary because the first match is always the oldest.